src/mlpack/methods/hmm/hmm.hpp

```diff
@@ -290,6 +290,73 @@ class HMM
    */
   double LogLikelihood(const arma::mat& dataSeq) const;
 
+  /**
+   * Compute the log of the scaling factor of the given emission probability
+   * at time t. To calculate the log-likelihood for the whole sequence,
+   * accumulate log scale over the entire sequence
+   *
+   * @param emissionLogProb emission probability at time t.
+   * probability up to time t-1
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log scale factor of the given sequence of emission at time t.
+   */
+  double LogScaleEmissionProb(const arma::vec& emissionLogProb,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
+  /**
+   * Compute the log-likelihood of the given emission probability up to time t
+   *
+   * @param emissionLogProb emission probability at time t.
+   * @param logLikelihood Log-likelihood of the given sequence of emission
+   * probability up to time t-1
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log-likelihood of the given sequence of emission up to time t.
+   */
+  double LogLikelihoodEmissionProb(const arma::vec& emissionLogProb,
+                       double &logLikelihood,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
```

Comment:
I'm not sure I understand what this function is for.  The documentation says `compute the log-likelihood of the given emission probability up to time t`, but I'm not sure from the documentation what the `given emission probability` is, since the user is passing a vector `emissionLogProb`.  I guess, I'm not sure how a user would use this function.  Could you clarify what the intent is?  Thanks! :+1:

Reply:
This is the one of the main points of suggesting this pull request at the first place. There are several reasons that one wants to calculate the emission probabilities by means other than GMM such as SVM, softmax regression, or NN. The advantages are faster and more accurate probability calculation and remarkably less memory footprint and CPU usage. GMM can become very slow and memory hug specially when dealing with high dimensional data let alone the assumptions that GMM imposed on the data. Thus this API is for this kind of scenario when one have the emission probabilities pre-calculated by a NN for example.

Reply:
I see!  I agree that is very useful.  It would be great if you could provide a quick example so that we can be sure we're on the same page.

But... one thing I wonder here, should we consider providing something higher-level?  So, e.g., the user can pass in a *model* that can compute log-probabilities for each emission, and then they pass in the original data points.  Anyway, just a thought.  We could totally leave that for later (or not at all).

We could probably also keep the function name simpler by having this just be another overload of `LogLikelihood()` (I think that will work).

Reply:
Regarding passing the model, I disagree because the model could be not from mlpack but from another library. For example one might use tensorflow or libsvm to calculate the probabilities, hence it's not possible to pass the model to the API.
I agree with the name. It's better to be another overload of `LogLikelihood()`

Reply:
I will add an example to the test unit

Reply:
> I agree with the name. It's better to be another overload of `LogLikelihood()`

Actually it cannot be called `LogLikelihood()` because it has the same signature as the original `LogLikelihood()`. Any suggestion?

Reply:
> Regarding passing the model, I disagree because the model could be not from mlpack but from another library. For example one might use tensorflow or libsvm to calculate the probabilities, hence it's not possible to pass the model to the API.

Good point!  I had not thought of that.

Reply:
> Actually it cannot be called LogLikelihood() because it has the same signature as another LogLikelihood(). Any suggestion?

Ah, I overlooked that.  Maybe `EmissionLogLikelihood()`?  I'm not sure that's too much better than what you have now though.